File: 112-path-sum/112-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool path_sum(TreeNode* root, int target) {
        if (!root)
            return false;
        
        if (!root->left and !root->right)
            return target == root->val;
        
        bool op1 = path_sum(root->left, target - root->val);
        bool op2 = path_sum(root->right, target - root->val);

        return (op1 or op2);
    }
    
    bool hasPathSum(TreeNode* root, int targetSum) {
        return path_sum(root, targetSum);
    }
};

File: 112-path-sum/112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def path_sum(self, root, targetSum) :
        if root == None:
            return False
        
        if root.left == None and root.right == None : 
            return targetSum == root.val
        
        op1 = self.path_sum(root.left, targetSum - root.val)
        op2 = self.path_sum(root.right, targetSum - root.val)
    
        return (op1 or op2)
    
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        return self.path_sum(root, targetSum)